src/H5TSrec_rwlock.c

```diff
@@ -0,0 +1,745 @@
+    FUNC_ENTER_PACKAGE_NAMECHECK_ONLY
+
+    assert(lock);
+    assert(H5TS_REC_RWLOCK_WRITE == lock->lock_type);
+    assert(lock->rec_write_lock_count >= 0);
+
+    lock->stats.write_locks_released++;
+
+    if (lock->rec_write_lock_count == 0)
+        lock->stats.real_write_locks_released++;
+
+    FUNC_LEAVE_NOAPI_VOID_NAMECHECK_ONLY
+} /* end H5TS__update_stats_wr_unlock() */
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS__rec_rwlock_get_stats
+ *
+ * Purpose:     Obtain a copy of the current statistics for a recursive
+ *              read / write lock.
+ *
+ * Note:        To obtain a consistent set of statistics, the function must
+ *              obtain the lock mutex.
+ *
+ * Return:      Non-negative on success / Negative on failure
+ *
+ *--------------------------------------------------------------------------
+ */
+herr_t
+H5TS__rec_rwlock_get_stats(H5TS_rec_rwlock_t *lock, H5TS_rec_rwlock_stats_t *stats)
+{
+    bool   have_mutex = false;
+    herr_t ret_value  = SUCCEED;
+
+    FUNC_ENTER_PACKAGE_NAMECHECK_ONLY
+
+    if (H5_UNLIKELY(NULL == lock || NULL == stats))
+        HGOTO_DONE(FAIL);
+
+    /* Acquire the mutex */
+    if (H5_UNLIKELY(H5TS_mutex_lock(&lock->mutex)))
+        HGOTO_DONE(FAIL);
+    have_mutex = true;
+
+    /* Copy R/W lock stats */
+    *stats = lock->stats;
+
+done:
+    if (H5_LIKELY(have_mutex))
+        if (H5_UNLIKELY(H5TS_mutex_unlock(&lock->mutex) < 0))
+            ret_value = FAIL;
+
+    FUNC_LEAVE_NOAPI_NAMECHECK_ONLY(ret_value)
+} /* end H5TS__rec_rwlock_get_stats() */
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS__rec_rwlock_reset_stats
+ *
+ * Purpose:     Reset the statistics for the supplied recursive read / write
+ *              lock.
+ *
+ * Note:        To obtain a consistent set of statistics, the function must
+ *              obtain the lock mutex.
+ *
+ * Return:      Non-negative on success / Negative on failure
+ *
+ *--------------------------------------------------------------------------
+ */
+herr_t
+H5TS__rec_rwlock_reset_stats(H5TS_rec_rwlock_t *lock)
+{
+    bool   have_mutex = false;
+    herr_t ret_value  = SUCCEED;
+
+    FUNC_ENTER_PACKAGE_NAMECHECK_ONLY
+
+    if (H5_UNLIKELY(NULL == lock))
+        HGOTO_DONE(FAIL);
+
+    /* Acquire the mutex */
+    if (H5_UNLIKELY(H5TS_mutex_lock(&lock->mutex) < 0))
+        HGOTO_DONE(FAIL);
+    have_mutex = true;
+
+    /* Reset stats */
+    memset(&lock->stats, 0, sizeof(lock->stats));
+
+done:
+    if (H5_LIKELY(have_mutex))
+        if (H5_UNLIKELY(H5TS_mutex_unlock(&lock->mutex) < 0))
+            ret_value = FAIL;
+
+    FUNC_LEAVE_NOAPI_NAMECHECK_ONLY(ret_value)
+} /* end H5TS__rec_rwlock_reset_stats() */
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS__rec_rwlock_print_stats
+ *
+ * Purpose:     Print the supplied pthresds recursive R/W lock statistics.
```

Comment:
Typo, though this isn't necessarily just for pthreads right?

Reply:
Yes, fixed